src/imas_coils.jl

```diff
@@ -1,13 +1,79 @@
 #= ==================================== =#
 #  IMAS.pf_active__coil to VacuumFields  #
 #= ==================================== =#
+
+"""
+    ElementCache{T}
+
+Cached derived quantities from IMAS coil data to avoid repeated allocation and computation.
+Each element in a coil has one ElementCache.
+Includes source geometry for cache validation.
+Mutable to enable in-place updates for zero-allocation cache invalidation.
+"""
+mutable struct ElementCache{T<:Real}
+    turns_with_sign::T
+    outline_r::Vector{T}
+    outline_z::Vector{T}
+    source_geometry::NamedTuple{(:r, :z, :width, :height), NTuple{4, T}}
+end
+
+"""
+    CurrentCache{T}
+
+Cached current_per_turn value for a specific time point.
+Stores the time when current was evaluated and its corresponding current_per_turn.
+"""
+struct CurrentCache{T<:Real}
+    time::T
+    current_per_turn::T
+end
+
 mutable struct GS_IMAS_pf_active__coil{T1<:Real,T2<:Real,T3<:Real,T4<:Real} <: AbstractSingleCoil{T1,T2,T3,T4}
     imas::IMAS.pf_active__coil{T1}
     tech::IMAS.build__pf_active__technology{T1}
     time0::Float64
     green_model::Symbol
+    _elements_cache::Vector{ElementCache{T1}}
+    _current_cache::CurrentCache{T1}
+
+    # Inner constructor (default initialization)
+    function GS_IMAS_pf_active__coil{T1,T2,T3,T4}(
+        imas::IMAS.pf_active__coil{T1},
+        tech::IMAS.build__pf_active__technology{T1},
+        time0::Float64,
+        green_model::Symbol) where {T1<:Real,T2<:Real,T3<:Real,T4<:Real}
+        empty_cache = ElementCache{T1}[]
+        invalid_current = CurrentCache{T1}(T1(NaN), T1(NaN))
+        new{T1,T2,T3,T4}(imas, tech, time0, green_model, empty_cache, invalid_current)
+    end
+
+    # Inner constructor (with element cache)
+    function GS_IMAS_pf_active__coil{T1,T2,T3,T4}(
+        imas::IMAS.pf_active__coil{T1},
+        tech::IMAS.build__pf_active__technology{T1},
+        time0::Float64,
+        green_model::Symbol,
+        elements_cache::Vector{ElementCache{T1}}) where {T1<:Real,T2<:Real,T3<:Real,T4<:Real}
+        invalid_current = CurrentCache{T1}(T1(NaN), T1(NaN))
+        new{T1,T2,T3,T4}(imas, tech, time0, green_model, elements_cache, invalid_current)
+    end
+
+    # Inner constructor (with element cache and current cache)
+    function GS_IMAS_pf_active__coil{T1,T2,T3,T4}(
+        imas::IMAS.pf_active__coil{T1},
+        tech::IMAS.build__pf_active__technology{T1},
+        time0::Float64,
+        green_model::Symbol,
+        elements_cache::Vector{ElementCache{T1}},
+        current_cache::CurrentCache{T1}) where {T1<:Real,T2<:Real,T3<:Real,T4<:Real}
+        new{T1,T2,T3,T4}(imas, tech, time0, green_model, elements_cache, current_cache)
+    end
```

Comment:
The first two inner constructors here should be outer constructors, since they change the number of arguments. The last one is just the default one, so won't be needed at all if the other ones are outer constructors. That said, you can simplify it all with [Base.@kwdef](https://docs.julialang.org/en/v1/base/base/#Base.@kwdef).